CorsixTH/Lua/entities/machine.lua

```diff
@@ -93,91 +96,130 @@ local function setSmoke(self, isSmoking)
+  if remaining_use_count < 4 then
     -- If the job of repairing the machine isn't queued, queue it now (higher priority)
-    if taskIndex == -1 then
+    if repair_task_dont_exist then
       local call = self.world.dispatcher:callForRepair(self, true, false, true)
       self.hospital:addHandymanTask(self, "repairing", 2, self.tile_x, self.tile_y, call)
       self.hospital:announceRepair(room)
     else -- Otherwise the task is already queued. Increase the priority to above that of machines with at least 4 uses left
        -- Upgrades task from low (1) priority to high (2) priority
        -- This does not lock the room, as happens when the task call starts at high priority
-      if self.hospital:getHandymanTaskPriority(taskIndex, "repairing") == 1 then
-        self.hospital:modifyHandymanTaskPriority(taskIndex, 2, "repairing")
+      if self.hospital:getHandymanTaskPriority(repair_task_index, "repairing") == 1 then
+        self.hospital:modifyHandymanTaskPriority(repair_task_index, 2, "repairing")
         self.hospital:announceRepair(room)
       end
     end
   -- Else if repair is needed, but not urgently
-  elseif threshold < 6 then
+  elseif remaining_use_count < 6 then
     -- If the job of repairing the machine isn't queued, queue it now (low priority)
-    if taskIndex == -1 then
+    if repair_task_dont_exist then
       local call = self.world.dispatcher:callForRepair(self)
       self.hospital:addHandymanTask(self, "repairing", 1, self.tile_x, self.tile_y, call)
     end
   end
+end
 
-  -- Update whether smoke gets displayed for this machine (and if so, how much)
-  self:calculateSmoke(room)
+--! Call on machine use. Handles crashing the machine
+function Machine:calculateIsMachineMustExplode(room, remaining_use_count)
+  -- Find a queued task for a handyman coming to repair this machine
+  local num_extinguishers = 0
+  local explosion_chance
+  local must_explode = false
+  -- Room is set to explode
+  if remaining_use_count < 1 then
+    -- If a fire extinguisher in the room, room has chance not to explode
+    -- Calculate count of extinguisher
+    for object, _ in pairs(room.objects) do
+      if object.object_type.id == "extinguisher" and num_extinguishers < 4 then
+        num_extinguishers = num_extinguishers + 1
+      end
+    end
+    if num_extinguishers == 0 or remaining_use_count < -3 then
+      -- If no enought extinguisher in room, or machine used 5 times over its strength always explode
+      must_explode = true
+    else
+      -- Explosion chance increases 20% with every use over strength, and reduced by 5% for every additional extinguisher (up to 3 extra) in the room bar the first one
+      explosion_chance = (2 / self.strength) + (remaining_use_count * -0.2) - (num_extinguishers * 0.05) + 0.05
+      -- Cap it until guaranteed explosion
+      explosion_chance = explosion_chance > 0.95 and 0.95 or explosion_chance
+      explosion_chance = explosion_chance < 0.05 and 0.05 or explosion_chance
```

Comment:
```suggestion
      explosion_chance = min(0.95, max(0.05, explosion_chance))
```

I don't understand line 189.

Reply:
> I don't understand line 189.

I am too